### PR TITLE
Automate the docker image builds

### LINKED ISSUE #174: Look into automating the Docker image builds
It may be possible to have a CI job that checks if there's been a dependency change (new package added, etc) and automatically rebuilds and pushes the new image to Dockerhub

### PR DESCRIPTION
**Pull request by joshessman-llnl**
This PR adds a GitHub action (currently runs manually on a user-selected branch) that builds and tests a Docker image, then pushes the image to Dockerhub.  It also saves the generated hostconfigs as artifacts that should be added to the repo.

The build time is currently around 1 hour - a fairly significant chunk is the MPI build required on the GCC 8 Ubuntu 16 image.

This has been tested on my fork of the serac repo: https://github.com/joshessman-llnl/serac/actions

Resolves #174

### CODE REVIEW

**Review by jamiebramwell — APPROVED** [review]
This is awesome! Thank for figuring it out!

File: src/docs/sphinx/docker_info.rst [hunk 1]
```diff
@@ -16,27 +16,19 @@ The following instructions apply to the creation of a new compiler image.
     #. Ensure that an Axom image exists on Dockerhub for the desired compiler.
```
Comment (white238) [inline src/docs/sphinx/docker_info.rst:17]: We should have a comment that most of the time you just need to use the Github Actions build the new docker images or clarifying that rebuilding existing docker images you dont need to follow these steps.

File: src/docs/sphinx/docker_info.rst [hunk 2]
```diff
@@ -16,27 +16,19 @@ The following instructions apply to the creation of a new compiler image.
...
+       for example, GCC 9.3 can be specified with ``./build_new_dockerfile.sh gcc 9.3``.  This will create a Dockerfile whose
+       name corresponds to a specific compiler, e.g., ``dockerfile_clang-10``.
+    #. Edit ``./github/workflows/docker_build_tpls.yml`` to add new job for the new compiler image.  This can be copy-pasted 
+       from one of the existing jobs - the only things that must be changed are the job name and ``TAG``, which should match the
+       name of the compiler/generated ``Dockerfile``.  For example, a build for ``dockerfile_clang-10`` must set ``TAG``
+       to ``clang-10``.  For clarity, the ``name`` field for the job should also be updated.
+    #. Commit and push the modified YML file and new Dockerfile, then go to the Actions tab on GitHub, select the "Docker TPL Build"
+       action, and run the workflow on the branch to which the above changes were pushed.  
+       **This will push new images to Dockerhub, overwriting existing images**.
+    #. Once the "Docker TPL Build" action completes, it will produce artifacts for each of the generated hostconfigs.  Download these
```
Comment (white238) [inline src/docs/sphinx/docker_info.rst]: nitpick to be consistent: hostconfigs -> host-configs

File: scripts/docker/dockerfile_gcc-8 [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
+SHELL ["/bin/bash", "-c"]
 RUN sudo apt-get update -y
-RUN sudo apt-get install curl ssh pkg-config build-essential libopenblas-dev mpich python3 -fy
+RUN sudo apt-get install curl wget ssh pkg-config build-essential libopenblas-dev tar python3 -fy
 RUN sudo apt-get -qq install -y --no-install-recommends gfortran-7 && sudo update-alternatives --install /usr/bin/gfortran gfortran /usr/bin/gfortran-7 100
 RUN sudo ln -f -s python3 /usr/bin/python # Compatibility with legacy Python
 
+# Avoid building CMake from scratch
+RUN wget https://cmake.org/files/v3.10/cmake-3.10.1-Linux-x86_64.sh
+RUN sudo sh cmake-3.10.1-Linux-x86_64.sh --prefix=/usr --exclude-subdir
```
Comment (white238) [inline scripts/docker/dockerfile_gcc-8:13]: :+1:

File: src/docs/sphinx/docker_info.rst [hunk 4]
```diff
@@ -16,27 +16,19 @@ The following instructions apply to the creation of a new compiler image.
...
-    #. Commit and push the Dockerfile to the Git repository.  To include the new image in CI jobs, add a new
-       ``matrix`` entry to ``azure-pipelines.yml``, modifying its attributes with the appropriate new image name and new
-       host-config file.
+    #. Go to the ``scripts/docker`` directory and run ``build_new_dockerfile.sh``, passing the compiler
+       name and version, e.g. for Clang 10, run ``./build_new_dockerfile.sh clang 10``.  Minor versions can also be specified,
+       for example, GCC 9.3 can be specified with ``./build_new_dockerfile.sh gcc 9.3``.  This will create a Dockerfile whose
+       name corresponds to a specific compiler, e.g., ``dockerfile_clang-10``.
+    #. Edit ``./github/workflows/docker_build_tpls.yml`` to add new job for the new compiler image.  This can be copy-pasted 
+       from one of the existing jobs - the only things that must be changed are the job name and ``TAG``, which should match the
+       name of the compiler/generated ``Dockerfile``.  For example, a build for ``dockerfile_clang-10`` must set ``TAG``
```
Comment (white238) [inline src/docs/sphinx/docker_info.rst:27]: There should be a comment in here that you may need to edit the generated dockerfile script for specific fixes. (like the many that are in docker_gcc-8)

File: scripts/docker/dockerfile_fake [hunk 5]
```diff
@@ -0,0 +1,16 @@
+FROM ubuntu:bionic
+RUN apt-get update && apt-get install -y sudo && rm -rf /var/lib/apt/lists/*
+RUN sudo useradd -m -s /bin/bash -G sudo serac
+
+WORKDIR "/home/serac"
+USER serac
```
Comment (white238) [inline scripts/docker/dockerfile_fake:10]: What is the purpose of this file? left over trial-and-error?

Comment (joshessman-llnl) [reply scripts/docker/dockerfile_fake:10]: Yeah, I thought it might be useful to check it in for future testing purposes

Comment (white238) [reply scripts/docker/dockerfile_fake:10]: Comment at the top then

**Review by white238 — APPROVED** [review]

**Comment by joshessman-llnl** [comment]
LGTM

**Pull request merged**